src/deep_approx_strict_equal.js

```diff
@@ -0,0 +1,118 @@
+import approxStrictEqual from "./approx_strict_equal";
+import epsilon from "./epsilon";
+
+/**
+ * Approximate strict equality for all elements of two structures, which may be
+ * null, undefined, number, string, array, or object.  This implements a subset
+ * of the Node sourc file 'lib/internal/util/comparisons.js', which is mirrored
+ * by 'internal/util/comparisons.js' in the CommonJS assert module.
+ *
+ * @param {number} actual The value to be tested.
+ * @param {number} expected The reference value.
+ * @param {number} tolerance The acceptable relative difference between elements.
+ * @return {boolean} Whether all numbers are within tolerance.
+ */
+function deepApproxStrictEqual(actual, expected, tolerance = epsilon) {
+    // Types must match
+    if (typeof expected !== typeof actual) {
+        return false;
+    }
+
+    // Undefined matches undefined
+    else if (expected === undefined) {
+        return actual === undefined;
+    }
+
+    // Null only equals null (check both sides to prevent attempts to access
+    // properties of null further down)
+    else if (expected === null) {
+        return actual === null;
+    }
+    else if (actual === null) {
+        return false;
+    }
+
+    // Booleans must be exactly equal
+    else if (typeof expected === "boolean") {
+        return expected === actual;
+    }
+
+    // Strings must be exactly equal
+    else if (typeof expected === "string") {
+        return expected === actual;
+    }
+
+    // Numbers must be approximately equal
+    else if (typeof expected === "number") {
+        return approxStrictEqual(actual, expected, tolerance);
+    }
+
+    // Corresponding array elements must be equal
+    else if (Array.isArray(expected)) {
+        return arrayApproxStrictEqual(actual, expected, tolerance);
+    }
+    else if (Array.isArray(actual)) {
+        return false;
+    }
+
+    // Objects must have matching keys and approximately equal values
+    else if (typeof expected === "object") {
+        return objectApproxStrictEqual(actual, expected, tolerance);
+    }
```

Comment:
I'm not sure about the utility of an object implementation here - it only really applies for objects that have numeric values and no prototype differences. Keeping this restricted to numbers and arrays of numbers seems to me like the most straightforward way to implement & document it, without getting into the rather sophisticated ways in which objects can be unequal.

Reply:
I had two use cases in mind:

1. A function returns multiple values in an object (e.g., `kMeansCluster`). However, this is rare and the parts can be checked for approximate equality one by one.

2. Complex numbers. Libraries like `complex.js` represent these as two-part objects; checking an array of such pairs without an object-handling case can be done with a couple of `map` calls to extract real and imaginary parts, but that feels clumsier than the top-level object decomposition in #1. On the other hand, complex values are much less common in statistics than in signal processing, so this could well be over-engineering.

Conclusion: strip out object handling, only do arrays, add objects back in later if people want them?

Reply:
> Conclusion: strip out object handling, only do arrays, add objects back in later if people want them?

👍 yep, sounds like the right approach. The kMeansCluster makes sense but is pretty rare, and I'm not sure that relative error on the components of a complex number is what folks desire/expect in that case.